google/cloud/forseti/common/util/email/base_email_connector.py

```diff
@@ -0,0 +1,121 @@
+        """Executes the sending of the email.
+
+        This needs to be a standalone method so that we can wrap it with retry,
+        and the final exception can be gracefully handled upstream.
+
+        Args:
+            email (mail): Connector mail object
+
+        Returns:
+            dict: urllib2 response
+        """
+        pass
+
+    @abc.abstractmethod
+    def send(self, email_sender=None, email_recipient=None,
+             email_subject=None, email_content=None, content_type=None,
+             attachment=None):
+        """Send an email.
+
+        This uses specific connector authentication details.
+
+        The minimum required info to send email are:
+        sender, recipient, subject, and content (the body)
+
+        Args:
+            email_sender (str): The email sender.
+            email_recipient (str): The email recipient.
+            email_subject (str): The email subject.
+            email_content (str): The email content (aka, body).
+            content_type (str): The email content type.
+            attachment (Attachment): An Email Connector Attachment.
+
+        Raises:
+            EmailSendError: An error with sending email has occurred.
+        """
+        pass
+
+    @classmethod
+    def render_from_template(cls, template_file, template_vars):
+        """Fill out an email template with template variables.
+
+        Args:
+            template_file (str): The location of email template in filesystem.
+            template_vars (dict): The template variables to fill into the
+                template.
+
+        Returns:
+            str: The template content, rendered with the provided variables.
+        """
+        template_searchpath = os.path.abspath(
+            os.path.join(os.path.dirname(__file__), '../../email_templates'))
+        template_loader = jinja2.FileSystemLoader(
+            searchpath=template_searchpath)
+        template_env = jinja2.Environment(loader=template_loader)
+        template = template_env.get_template(template_file)
+        return template.render(template_vars)
+
+    @classmethod
+    def create_attachment(cls, file_location, content_type, filename,
+                          disposition='attachment', content_id=None):
+        """Create a SendGrid attachment.
+
+        Email connector attachments file content must be base64 encoded.
+
+        Args:
+            file_location (str): The path of the file.
+            content_type (str): The content type of the attachment.
+            filename (str): The filename of attachment.
+            disposition (str): Content disposition, defaults to "attachment".
+            content_id (str): The content id.
+
+        Returns:
+            Attachment: A Connector Attachment.
+        """
+        file_content = ''
+        with open(file_location, 'rb') as f:
+            file_content = f.read()
+        content = base64.b64encode(file_content)
+
+        attachment = mail.Attachment()
```

Comment:
You are using the sendgrid attachment here in the base connector?  We should not have a provider-specific connector here, right?

Reply:
I see. I will move this to sendgrid_connector.py